R/derive_extreme_records.R

```diff
@@ -0,0 +1,157 @@
+#'
+#'   *Permitted Values:* list of variables created by `vars()`
+#'
+#' @param order Sort order
+#'
+#'   Within each by group the observations are ordered by the specified order.
+#'
+#'   *Permitted Values:* list of variables or `desc(<variable>)` function calls
+#'   created by `vars()`, e.g., `vars(ADT, desc(AVAL))`
+#'
+#' @param mode Selection mode (first or last)
+#'
+#'   If `"first"` is specified, the first observation of each by group is added
+#'   to the input dataset. If `"last"` is specified, the last observation of
+#'   each by group is added to the input dataset.
+#'
+#'   *Permitted Values:* `"first"`, `"last"`
+#'
+#' @param check_type Check uniqueness?
+#'
+#'   If `"warning"` or `"error"` is specified, the specified message is issued
+#'   if the observations of the filtered input dataset are not unique with
+#'   respect to the by variables and the order.
+#'
+#'   *Default:* `"warning"`
+#'
+#'   *Permitted Values:* `"none"`, `"warning"`, `"error"`
+#'
+#' @param filter Filter for observations to consider
+#'
+#'   Only observations fulfilling the specified condition are taken into account
+#'   for selecting the first or last observation. If the parameter is not
+#'   specified, all observations are considered.
+#'
+#'   *Default*: `NULL`
+#'
+#'   *Permitted Values*: a condition
+#'
+#' @param set_values_to Variables to be set
+#'
+#'   The specified variables are set to the specified values for the new
+#'   observations. For example `vars(DTYPE = "MINIMUM")` defines the derivation
+#'   type for the new observations.
+#'
+#'   *Permitted Values:* List of variable-value pairs
+#'
+#' @details
+#'   1. The input dataset is restricted as specified by the `filter` parameter.
+#'   1. For each group (with respect to the variables specified for the
+#'   `by_vars` parameter) the first or last observation (with respect to the
+#'   order specified for the `order` parameter and the mode specified for the
+#'   `mode` parameter) is selected.
+#'   1. The variables specified by the `set_values_to` parameter are added to
+#'   the selected observations.
+#'   1. The observations are added to input dataset.
+#'
+#' @author Stefan Bundfuss
+#'
+#' @return The input dataset with the first or last observation of each by group
+#'   added as new observations.
+#'
+#' @keywords derivation bds
+#'
+#' @export
+#'
+#' @examples
+#' adlb <- tibble::tribble(
+#'   ~USUBJID, ~AVISITN, ~AVAL, ~LBSEQ,
+#'   "1",      1,        113,   1,
+#'   "1",      2,        113,   2,
+#'   "1",      3,        117,   3,
+#'   "2",      1,        101,   1,
+#'   "2",      2,        101,   2,
+#'   "2",      3,         95,   3
+#' )
+#'
+#' # Add a new visit for the minium value (first observation if not unique)
+#' derive_extreme_records(
+#'   adlb,
+#'   by_vars = vars(USUBJID),
+#'   order = vars(AVAL, AVISITN),
+#'   mode = "first",
+#'   set_values_to = vars(
+#'     AVISITN = 97,
+#'     DTYPE = "MINIMUM"
+#'   )
```

Comment:
Would add an example with filter = !is.na(AVAL) to inform user maybe to pay attention to missing values?

Reply:
I have added `filter = !is.na(AVAL)` to the minimum and the maximum example.